Have `help` search DataTypes with their short names (helps with #11438)

fcard opened the pull request.

fcard committed:
Fix `help` searching for DataTypes

+ `apropos` fix

vtjnash commented on test/help.jl:
```diff
@@ -0,0 +1,5 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+for testcase in [Dates.DateTime, RoundingMode, Libc.TmStruct]
+    @test !startswith(sprint(help, testcase), "DataType   :")
```
there needs to be a positive assertion test also

vtjnash commented on test/help.jl:
```diff
@@ -0,0 +1,5 @@
+
+for testcase in [Dates.DateTime, RoundingMode, Libc.TmStruct]
+    @test !startswith(sprint(help, testcase), "DataType   :")
+end
```
i think this test can go in the `docs.jl` test file

vtjnash commented:
would it be possible to change the lookup dict to reference the object itself as the key instead of a string? would that help solve this?

fcard commented:
I suppose one could parse+eval the object names as they are read from the database. All objects would need to be guaranteed to be defined by then. However, macros would still need to be referenced by string or symbol.

Is there anything wrong with the current approach?

vtjnash commented:
i was considering the behavior of the following case:
```julia
module Trees
  @doc "rough" hasbark(::Tree) = true
end
module Animals
  @doc "ruff" hasbark(::Dog) = true
  @doc "fur" Main.Trees.hasbark(::Dog) = false
end
```
```julia
@doc Dog.hasbark
# do we want doc for Tree.hasbark to show up here?
```

but now that i think about it, it may actually be true that printing documentation of all instances of indexed under `hasbark` may actually be preferable (but with some annotation to hint to the user that they are not the same function?)

fcard commented:
Well, `@doc`s dictionary is not the same as `help`s, and `@doc`s already maps from Objects.

Even in the case of `help`, it only shows all functions from all modules if only the function name is provided, in string form.

I made a little script that should showcase all behaviors: https://gist.github.com/fcard/752cb2afee14e18615c4

However, it's true that with types, sending the object to help will show documentation from all modules (maybe that's what you meant and used `@doc` and functions because it was easier...?). I think I know to fix that without changing the `string->doc` model, though. Given that the functionality of retrieving all documentation from all modules is already present with `help("TypeName")`, I think I will do just that. Any objections?

Also, should I do something about the merge conflict? I looked everywhere how solve it myself but apparently the mergers have to do it themselves.

tkelman commented:
You need to do a local clone of this branch, rebase, then force push back to your fork on github.

fcard commented:
@tkelman Cool! Thanks, I had thought to do similar but with merge, but for some reason was scared that all the new commits from master would be referenced here and they would flood the page. Also I just reread `CONTRIBUTING.md` and noticed that it actually explains this, so my bad, somehow I missed this the others times I read it.

-

Anyway, I added a new clause such that the documentation of a datatype will be printed, given its name and module, only if the datatype is defined/imported in a module path (optionally omitting "Base.") that has documentation associated with its name. e.g.:
  * `Base.Set` has documentation associated with it in "Base", and its defined there.
  * `Core.Array` has documentation associated with it in "Base", and its imported there.
  * `Base.Dates.DateTime` has documentation associated with it in "Dates", and it's defined in "Base.Dates"
  * `Base.Rounding.RoundingMode` has documentation defined for "Base" and it's imported there.

I think it covers all types in the database, and will continue to if they are written in the same style.

I added a few tests to check this new clause, but the module evals and global variable manipulation is making me uneasy. If you want me to remove them, I will do it. (dunno how else to test it, though)

tkelman commented on base/help.jl:
```diff
@@ -147,7 +166,7 @@ apropos(s::AbstractString) = apropos(STDOUT, s)
     init_help()
     n = 0
-    r = Regex("\\Q$txt", Base.PCRE.CASELESS)
+    r = Regex("\\Q$txt", "i")
```
would a test have caught this? can you add one?

fcard commented:
The `Regex` was throwing a method error. Should I just use the `apropos` function somewhere and let the errors explain themselves? Or should I `catch` this error specifically and write a message detailing this case?

tkelman commented:
But you've fixed that method error now, right? I would call `apropos` on something simple, redirecting its output into an `IOBuffer`, then check that the contents of that `IOBuffer` match what you expected. If we try to run that process on master I'm guessing it would fail because of the regex issue - probably fallout from upgrading to PCRE2, but there was no test coverage of that line so the problem was missed until now. If you add a test that exercises that line of code, then we can hopefully avoid breaking it again in the future.

fcard commented:
Yeah, that is what I was thinking more or less. I thought that catching the error would let me explain the issue better, but rereading the error message, it's descriptive enough. Also, this covers more possible errors.

One thing about this `Regex` method change is that there is a designated default value for the new argument (a constant called `DEFAULT_MATCH_OPTS`) but the method doesn't use it. Maybe using that default would save the hunting for other similar fallouts.

tkelman commented:
> One thing about this `Regex` method change is that there is a designated default value for the new argument (a constant called `DEFAULT_MATCH_OPTS`) but the method doesn't use it. Maybe using that default would save the hunting for other similar fallouts.

I don't quite understand what you mean here. cc @malmaud who did the PCRE2 work recently.

I like your test though, nice and concise and indeed it fails on master.

fcard commented:
Before, `Regex` had a method `(String, Integer)`, the string to be tested and "compiler options". That method now takes a new argument, "match options". All uses of `Regex` that used the `(String, Integer)` form are now broken. However, there is a designated default value for that new argument, so the method `(String, Integer)` could be brought back and solve the problem.

```julia
Regex(pattern::AbstractString, compiler_options::Integer) =
    Regex(pattern, compiler_opts, DEFAULT_MATCH_OPTS)
```

Did I explain better this time?

tkelman commented:
Okay, thanks, that makes more sense, but is sounding like a separate issue that's unrelated to your changes to `help` here, right?

fcard commented:
Yes, it was just a quick thing I wanted to point out, didn't mean for it to drag on :P

tkelman commented:
No, I'm glad you did, I think it might even be worth opening a separate issue on so we don't forget about it here. Or at the very least leaving a comment on the merged PCRE2 PR to continue discussion there. That could have regex compatibility consequences in packages and elsewhere.

fcard commented:
Ok, issue opened! Anything else left for me to do here?

tkelman commented:
@one-more-minute @vtjnash do you think this can be merged as-is?

one-more-minute commented:
Looks like a nice change and implementation to me.

(@fcard You probably know this already but just FYI, the plan is to move away from and deprecate the current help system over the next releases. The improvements are still welcome, of course.)

fcard commented:
@one-more-minute I assumed as much, but hopefully it will be useful in the while it's still relevant. Initially it was also a 1/2 line fix so I thought "I may as well, even if it's temporary", and when enhancements were proposed I took it as a good opportunity to practice and learn.

tkelman commented:
@one-more-minute I'll take that as a "yes we should merge this" - any reason you didn't?

fcard commented:
Thanks for merging! I guess this closes #11690, right?
